captum/attr/_utils/interpretable_input.py

```diff
@@ -0,0 +1,100 @@
+    Captum understand how to perturb various inputs.
+
+    The concept Interpretable Input mainly comes from the following two papers:
+
+    `"Why Should I Trust You?": Explaining the Predictions of Any Classifier
+    <https://arxiv.org/abs/1602.04938>`_
+
+    `A Unified Approach to Interpreting Model Predictions
+    <https://arxiv.org/abs/1705.07874>`_
+
+    which is also referred to as interpretable representation or simplified
+    input. It can be represented as a mapping function:
+
+    .. math::
+        x = h_x(x')
+
+    where :math:`x` is the model input, which can be anything that the model consumes;
+    :math:`x'` is the interpretable input used in the attribution algorithms
+    (it must be a PyTorch tensor in Captum), which is often
+    binary indicating the “presence” or “absence”; :math:`h_x` is the
+    transformer. It is supposed to work with perturbation-based attribution methods,
+    but if :math:`h_x` is differentiable, it may also be used
+    in gradient-based methods.
+
+    InterpretableInput is the abstarct class defining the interface. Captum provides
+    the child implementations for some common input formats,
+    like text and sparse features. Users can inherit this
+    class to create other types of customized input.
+
+    (We expect to support InterpretableInput in all attribution methods, but it
+    only allowed in certain attribution classes like LLMAttribution for now.)
+    """
+
+    def to_tensor(self) -> Tensor:
+        """
+        Return the interpretable representation of this input as a tensor
+
+        Returns:
+
+            itp_tensor (Tensor): interpretable tensor
+        """
+        pass
+
+    def to_model_input(self, itp_tensor: Optional[Tensor] = None) -> Any:
+        """
+        Get the the (perturbed) input in the format required by the model
+        based on the given (perturbed) interpretable representation.
+
+        Args:
+
+            itp_tensor (Tensor, optional): tensor of the interpretable representation
+                    of this input. If it is None, assume the interpretable
+                    representation is pristine and return the orignal model input
+                    Default: None.
+
+
+        Returns:
+
+            model_input (Any): model input passed to the forward function
+        """
+        pass
+
+    def format_attr(self, itp_attr) -> Tensor:
+        """
+        Format the attribution of the intepretable feature if needed.
+        The way of formatting depends on the specific interpretable input type.
+        A common use is if the interpretable features are the mask groups of the raw
+        input elements, the attribution of the interpretable features can be scattered
+        back to the model input shape.
+
+        Args:
+
+                itp_attr (Tensor): attributions of the interpretable features
+
+        Returns:
+
+                attr (Tensor): formatted attribution
```

Comment:
Would it make sense to allow the return type to be Any here to allow a custom dict, etc. to represent the attributions in a custom format similar to input rather than a tensor?

Reply:
Maybe! My only concern is some of our downstream tasks assume the attr is tensor, like stats aggregation. We can discuss this when we face an Input type that using tensor as attr will be very confusing or inconvenient